crates/sui-framework/packages/sui-framework/sources/token.move

```diff
@@ -0,0 +1,727 @@
+    public fun confirm_request_mut<T>(
+        policy: &mut TokenPolicy<T>,
+        request: ActionRequest<T>,
+        ctx: &mut TxContext
+    ): (String, u64, address, Option<address>) {
+        assert!(vec_map::contains(&policy.rules, &request.name), EUnknownAction);
+        if (option::is_some(&request.spent_balance)) {
+            balance::join(
+                &mut policy.spent_balance,
+                option::extract(&mut request.spent_balance)
+            );
+        };
+
+        confirm_request(policy, request, ctx)
+    }
+
+    /// Confirm an `ActionRequest` as the `TokenPolicyCap` owner. This function
+    /// allows `TokenPolicy` owner to perform Capability-gated actions ignoring
+    /// the ruleset specified in the `TokenPolicy`.
+    ///
+    /// Aborts if request contains `spent_balance` due to inability of the
+    /// `TokenPolicyCap` to decrease supply. For scenarios like this a
+    /// `TreasuryCap` is required (see `confirm_with_treasury_cap`).
+    public fun confirm_with_policy_cap<T>(
+        _policy_cap: &TokenPolicyCap<T>,
+        request: ActionRequest<T>,
+        _ctx: &mut TxContext
+    ): (String, u64, address, Option<address>) {
+        assert!(option::is_none(&request.spent_balance), ECantConsumeBalance);
+
+        let ActionRequest {
+            name, amount, sender, recipient, approvals: _, spent_balance
+        } = request;
+
+        option::destroy_none(spent_balance);
+
+        (name, amount, sender, recipient)
+    }
+
+    /// Confirm an `ActionRequest` as the `TreasuryCap` owner. This function
+    /// allows `TreasuryCap` owner to perform Capability-gated actions ignoring
+    /// the ruleset specified in the `TokenPolicy`.
+    ///
+    /// Unlike `confirm_with_policy_cap` this function allows `spent_balance`
+    /// to be consumed, decreasing the `total_supply` of the `Token`.
+    public fun confirm_with_treasury_cap<T>(
+        treasury_cap: &mut TreasuryCap<T>,
+        request: ActionRequest<T>,
+        _ctx: &mut TxContext
+    ): (String, u64, address, Option<address>) {
+        let ActionRequest {
+            name, amount, sender, recipient, approvals: _,
+            spent_balance
+        } = request;
+
+        if (option::is_some(&spent_balance)) {
+            balance::decrease_supply(
+                coin::supply_mut(treasury_cap),
+                option::destroy_some(spent_balance)
+            );
+        } else {
+            option::destroy_none(spent_balance);
+        };
+
+        (name, amount, sender, recipient)
+    }
+
+    // === Rules API ===
+
+    /// Add an "approval" to the `ActionRequest` by providing a Witness.
+    /// Intended to be used by Rules to add their own approvals, however, can
+    /// be used to add arbitrary approvals to the request (not only the ones
+    /// required by the `TokenPolicy`).
+    public fun add_approval<T, W: drop>(
+        _t: W, request: &mut ActionRequest<T>, _ctx: &mut TxContext
```

Comment:
We may want `request` to be the first argument to make this more amenable to receiver syntax

Reply:
Does it apply to other functions as well? The issue here is that we always put witness first, and it's already this way in a lot of packages / projects.

Reply:
I would say it's worth taking an eye to all these functions with receiver syntax in-mind (although these were the ones that popped out the most to me). 

Might be worth bucking the trend of the existing ones since before we didn't really have that good of a reason whereas now there's more of a reason to change the order of these arguments possibly? 

I don't have super strong opinions here though and whether the order should be changed since there's definitely an argument to be made for consistency as well.